connectors/bitbucket-server/src/main/java/com/vmware/connectors/bitbucket/server/BitBucketServerController.java

```diff
@@ -0,0 +1,435 @@
+                                                @RequestHeader(AUTH_HEADER) final String authHeader,
+                                                @RequestHeader(BASE_URL_HEADER) final String baseUrl) {
+
+        logger.debug("Decline action for bitbucket server pull request: {}, baseURL: {}",
+                pullRequest,
+                baseUrl);
+
+        return performBitBucketServerAction(baseUrl, authHeader, pullRequest, BitBucketServerAction.DECLINE);
+    }
+
+    @PostMapping(
+            path = "/api/v1/{projectKey}/{repositorySlug}/{pullRequestId}/comments",
+            consumes = MediaType.APPLICATION_FORM_URLENCODED_VALUE
+    )
+    public Single<ResponseEntity<String>> comments(final BitBucketServerPullRequest pullRequest,
+                                                  @RequestParam(BitBucketServerConstants.COMMENT_PARAM_KEY) final String comment,
+                                                  @RequestHeader(AUTH_HEADER) final String authHeader,
+                                                  @RequestHeader(BASE_URL_HEADER) final String baseUrl) {
+
+        logger.debug("Comment action for bitbucket server pull request: {}, baseURL: {}",
+                pullRequest,
+                baseUrl);
+
+        final HttpHeaders headers = buildHeaders(authHeader);
+        headers.add(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE);
+
+        final String url = String.format(BitBucketServerConstants.BITBUCKET_PR_COMMENT_URL_FORMAT,
+                baseUrl,
+                pullRequest.getProjectKey(),
+                pullRequest.getRepositorySlug(),
+                pullRequest.getPullRequestId());
+
+        final BitBucketServerComment bitBucketServerComment = new BitBucketServerComment(comment);
+
+        final ListenableFuture<ResponseEntity<String>> response = this.rest.exchange(
+                url,
+                HttpMethod.POST,
+                new HttpEntity<>(bitBucketServerComment, headers),
+                String.class);
+        return Async.toSingle(response);
+    }
+
+    private List<String> getComments(final String baseUrl,
+                                     final HttpHeaders headers,
+                                     final BitBucketServerPullRequest pullRequest) {
+        final String url = String.format(BitBucketServerConstants.BITBUCKET_ACTIVITIES_URL_FORMAT,
+                baseUrl,
+                pullRequest.getProjectKey(),
+                pullRequest.getRepositorySlug(),
+                pullRequest.getPullRequestId());
+
+        final ListenableFuture<ResponseEntity<JsonDocument>> response = this.rest.exchange(
+                url,
+                HttpMethod.GET,
+                new HttpEntity<>(headers),
+                JsonDocument.class);
+
+        final JsonDocument[] jsonDocuments = new JsonDocument[1];
+        Async.toSingle(response)
+                .subscribe(entity -> {
```

Comment:
This is still broken. You never want to call `subscribe` in a connector. Spring does that for us further up the call stack . In the real, live connector, line 193 is going to be called before line 191.

You are getting comments asynchronously so it stands to reason that this method _cannot_ return `List<String>`. Instead it is going to have to return an Rx object, which will be incorporated into another RxJava pipeline.

Do let me know if you get stuck on this and I will suggest a solution.

Reply:
I have done the changes to remove subscribe. I have not addressed all the pull requests comments. I have pushed only some of the changes. Apologies for the inconvenience caused.

Reply:
I have done the changes per @RyanBard suggestion.